cranelift/codegen/src/verifier/mod.rs

```diff
@@ -1124,6 +1148,66 @@ impl<'a> Verifier<'a> {
         }
     }
 
+    /// Checks If a `load` or `store` access is within `heap_addr` bounds.
+    fn verify_heap_addr_bounds(
+        &self,
+        inst: Inst,
+        opcode: Opcode,
+        addr: Value,
+        offset: Offset32,
+        errors: &mut VerifierErrors,
+    ) -> VerifierStepResult<()> {
+        let dfg = &self.func.dfg;
+
+        // First, let's check if `addr` comes from a `heap_addr` instruction.
+        let valid_size = match dfg.value_def(addr) {
+            ValueDef::Result(def_inst, _) => match dfg[def_inst] {
+                ir::InstructionData::HeapAddr { imm, .. } => u32::from(imm) as i64,
+                // Not a `heap_addr`, we can't verify this.
```

Comment:
In principle I suppose we could verify arithmetic where one operand is a `heap_addr` and the rest are all constants. I'm not suggesting that should be in this PR (unless you really want to), but maybe note it in this comment?